Add eventDocs & extrinsicDocs query params to /blocks

emostov opened the pull request:
Closes: #257 

This PR adds 2 optional query params for the `/blocks` endpoint: `eventDocs` & `extrinsicDocs`. They are both boolean parameters that allow the user to opt into having `docs` in the response. `eventDocs=true` will include `docs` for every event while `extrinsicDocs` will include `docs` for top level extrinsics int the `extrinsics` array/

See sample response from `/blocks/1470716?eventDocs=true&extrinsicDocs=true` here: https://gist.github.com/emostov/dafb0b017df13b64fb411f6e6d769469

joepetrowski commented on src/controllers/blocks/BlocksController.ts:
```diff
@@ -13,6 +13,15 @@ import AbstractController from '../AbstractController';
+ * Query:
+ * - (Optional) `eventDocs`: When set to `true`, every event will have an extra
+ * 	`docs` property containing an array of strings that represent each line of
+ * 	events documentation
```
```suggestion
 * 	the event's documentation.
```

joepetrowski commented on src/controllers/blocks/BlocksController.ts:
```diff
@@ -13,6 +13,15 @@ import AbstractController from '../AbstractController';
+ * 	events documentation
+ * - (Optional) `extrinsicDocs`: When set to `true`, every extrinsic will have an extra
+ * 	`docs` property containing an array of strings that represent each line of
+ * 	events documentation
```
```suggestion
 * 	the extrinsic's documentation.
```

dvdplm commented on src/services/blocks/BlocksService.ts:
```diff
@@ -212,6 +229,7 @@ export class BlocksService extends AbstractService {
 				success: defaultSuccess,
 				// paysFee overrides to bool if `system.ExtrinsicSuccess|ExtrinsicFailed` event is present
 				paysFee: null as null | boolean,
+				docs: extrinsicDocs ? extrinsic.meta.documentation : undefined,
```
Elsewhere you have `const eventDocsArg = eventDocs === 'true';` which implies that this should be a `bool`, so I'd expect this to be `docs: extrinsicDocs ? extrinsic.meta.documentation : false,`. Or why do you prefer `undefined` here?

kianenigma replied:
AFAIU This is setting the _final value_ itself inside a `{}`, thus the field `docs` will be set either to `extrinsic.meta.documentation` or `undefined`.

emostov replied:
The reason I chose `undefined` here is that when a property (in this case `docs`) is set to undefined in a response body (in this case a plain-old-javascript-object), the response body will omit that field. If we wanted to keep the field in even if they did not want docs I think `null` would make the most sense.

I think have `docs` set to `false` when the user does not opt in to them might be confusing because the inverse would be having an array of strings for `docs`, not `true`.

dvdplm commented on src/controllers/blocks/BlocksController.ts:
```diff
@@ -13,6 +13,15 @@ import AbstractController from '../AbstractController';
+ * 	`docs` property containing an array of strings that represent each line of
+ * 	events documentation
+ * - (Optional) `extrinsicDocs`: When set to `true`, every extrinsic will have an extra
+ * 	`docs` property containing an array of strings that represent each line of
```
I think it's unfortunate that the docs come back as an array of strings rather than a single string (with newlines). Is that something that is easy to change?

kianenigma replied:
Yeah I guess we could do some parsing.. The array of string is the side effect of how the metadata is being generated.

emostov replied:
We certainly can do parsing, appending a new line char and then joining them with a string. I will make that change and update the sample response so we can look at it.

mattrutherford commented on src/controllers/blocks/BlocksController.ts:
```diff
@@ -73,12 +82,18 @@ export default class BlocksController extends AbstractController<
 		const hash = await this.api.rpc.chain.getFinalizedHead();
 
+		const eventDocsArg = eventDocs === 'true';
+		const extrsinsicDocsArg = extrinsicDocs === 'true';
```
small typo here maybe: `extrsinsicDocsArg`

kianenigma commented on src/services/test-helpers/responses/blocks/blocks789629.json:
```diff
@@ -52,11 +52,34 @@
       ],
       "success": true,
-      "paysFee": true
+      "paysFee": true,
```
what's up with this file?

emostov replied:
Is used to test the response from the service. This way anytime there is a user contract change in the service it forces one to change this file and document the change. Additionally it helps prevent accidental regressions in the user contract.

kianenigma commented on src/types/responses/FrameMethod.ts:
```diff
@@ -0,0 +1,11 @@
+	method: string;
+}
+
+export function isFrameMethod(thing: unknown): thing is IFrameMethod {
+	return (
+		typeof (thing as IFrameMethod).pallet === 'string' &&
+		typeof (thing as IFrameMethod).method === 'string'
+	);
+}
```
fascinating 🌈 :D

kianenigma reviewed (approved):
LGTM

emostov commented:
I have refactored and changed the response so that instead of an array of strings, the docs is not a single string with new line characters at the end of each line. Take a look at the updated test respons json and the gist to see what the response looks like.
